src/CheatingStitchingTool.cc

```diff
@@ -0,0 +1,140 @@
+/**
+ *  @file   src/CheatingStitching.cc
+ *
+ *  @brief  Implementation of the cheating stitching tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include <larpandoracontent/LArHelpers/LArMCParticleHelper.h>
+#include <larpandoracontent/LArHelpers/LArPfoHelper.h>
+
+#include "CheatingStitchingTool.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+CheatingStitchingTool::CheatingStitchingTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void CheatingStitchingTool::Run(const MasterAlgorithm *const pAlgorithm, const PfoList *const /*pMultiPfoList*/, PfoToLArTPCMap &pfoToLArTPCMap,
+    PfoToFloatMap &/*stitchedPfosToX0Map*/)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    if (this->GetPandora().GetGeometry()->GetLArTPCMap().size() < 2)
+        return;
+
+    if (pfoToLArTPCMap.empty())
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+    const MCParticleList *pMCParticleList{nullptr};
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*pAlgorithm, pMCParticleList));
+    const PfoList *pPfoList{nullptr};
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*pAlgorithm, pPfoList));
+
+    std::map<const MCParticle *, std::map<unsigned int, std::vector<const ParticleFlowObject *>>> mcParticleToLArTPCToPfosMap;
+
+    // Find all PFOs associated to each MCParticle
+    for (const ParticleFlowObject *const pPfo : *pPfoList)
+    {
+        CaloHitList allHits;
+        LArPfoHelper::GetAllCaloHits(pPfo, allHits);
+        std::map<const MCParticle *, unsigned int> mcParticleToHitCountMap;
+
+        for (const CaloHit *const pCaloHit : allHits)
+        {
+            MCParticleVector mcParticleVector;
+            for (const auto &weightMapEntry : pCaloHit->GetMCParticleWeightMap())
+                mcParticleVector.push_back(weightMapEntry.first);
+            std::sort(mcParticleVector.begin(), mcParticleVector.end(), LArMCParticleHelper::SortByMomentum);
+
+            // Consider only the primary MCParticle associated to this hit
+            if (!mcParticleVector.empty())
+                ++mcParticleToHitCountMap[mcParticleVector.front()];
+        }
+
+        // Find the MCParticle with the most hits associated to this PFO
+        const MCParticle *pMCParticle{nullptr};
```

Comment:
Is there some SDK functionality that can be used here (e.g. variations of GetMainMCParticle) to make it a bit slimmer?